### ISSUE TITLE
Webpage constantly refreshing after update to app

### ISSUE THREAD

**Issue by elie222** [meteor/meteor#7115]
Hi,

When I update my production app and a user visits the site, the site will constantly load and refresh until the user clicks the stop button (and perhaps a refresh after that). This only happens when I redeploy my app and will only happen the first time the user visits the site.

I'm running in production on Meteor 1.3.2.4. Using mupx to deploy. Using nginx as a reverse proxy load balancer.

I do not have a way to reproduce this in a local repo, but this issue happens every time I update my app and others are reporting the same issue.

This problem did not happen in Meteor 1.2 with the same nginx configuration.

This has been brought up in the forums and a few people have this issue, but no solution has yet to be found: https://forums.meteor.com/t/app-constantly-refreshing-after-an-update/23586

**Comment by laosb** [comment]
As I stated after the forum post, it must be something related to cache. Maybe `appcache`, maybe nginx cache. I'm not sure. You can disable hot reload by `meteor remove autoupdate` for now.

**Comment by JesperWe** [comment]
Confirming this for a site behind nginx but using no caching. Does not only happen on hot code pushes, happens when a new user enters the site as well, so remove autoupdate is not a solution.

**Comment by abernix** [comment]
Let's continue to debug this in the forum post (just one place) for the time-being until we can figure out if it's an actual bug or an Nginx configuration and work to get a reproduction made.  I just made a post there.

**Comment by rjakobsson** [comment]
I'm getting the same behavior for _some_ clients, not all. Hosting on Heruko.

**Comment by JesperWe** [comment]
@rjakobsson please add your info in the forum post (see above), this is hard to debug because it is intermittent, but please read the thread and see if you can spot any similarities to your case.

**Comment by betacar** [comment]
Can confirm it still happens in 1.3.1.

**Comment by CharlesBergmeier** [comment]
When can we expect this issue to be resolved?  This issue has been open for almost a month now with no resolution, and represents a serious bug that renders the service unusable.

**Comment by JesperWe** [comment]
@CharlesBergmeier please read my post at https://forums.meteor.com/t/app-constantly-refreshing-after-an-update/23586/73, and see if you can contribute any additional info...

**Comment by elie222** [comment]
I agree with Charles. It really is making the service unusable to the point of clients inadvertently DDOSing  my servers for a week at a time. I'd be happy to give more info linking to deployments if someone at Meteor is going to look into this.

**Comment by jrrera** [comment]
We were experiencing this issue as well. We were able to solve it by removing nginx caching. In particular, this line:

```
if ($uri != '/') {		
  expires 30d;		
}
```

**Comment by laosb** [comment]
@CharlesBergmeier @elie222 I'm sorry if this bug annoys you, but we didn't have enough information to figure out where the actual problem is and what's the best solution is. Please goto the forum post above and provide more information so we can set off working on this.

**Comment by elie222** [comment]
Yeah. I realise no one has given any reproducible test case. I can't even
reproduce it on my own app anymore although it used to happen consistently.

The problem with discussing it in the forum is that it doesn't seem that
many MDG people are in the discussion. I think one was before he went on
holiday. And it feels a little like we're stabbing in the dark trying to
guess at what's happening without knowing the meteor internals too well.

**Comment by JesperWe** [comment]
@abernix is not MDG Staff, he is a knowledgeable community member who has taken the task of helping out in the forums.

**Comment by elie222** [comment]
Ah! My bad! Then it really would be nice to have an MDG member with some
input!

**Comment by laosb** [comment]
@elie222 You can tell MDG members apart from other community member by the badge on the top right corner:)

**Comment by elie222** [comment]
Good to know. Thanks

**Comment by cyllu2** [comment]
Is this the same issue that I'm getting in production, when sometimes (1 out of 100 perhaps) after a hot reload the page will start to refresh repeatedly and I have to restart the server to continue?

You guys are talking about Nginx and stuff, so I'm not sure if you're discussing a different issue and if the one I'm having will carry over to deployment.

**Comment by btoueg** [comment]
I am impacted by this problem.

Using Meteor 1.3.5.1 behind nginx.

**Comment by rjakobsson** [comment]
The current solution was to shut down client side reloading upon new version. See the forum thread mentioned earlier @btoueg .

**Comment by jaskinn** [comment]
So I assume @rjakobsson that you mean the comment by the meteor user 'pauldulong'

> When I disabled the Hot Code Push (HCP) on the client side Meteor._reload.onMigrate(function() {return
> [false];}); by overwriting the reload function, the problem seems to be solved.

**Comment by rjakobsson** [comment]
Sorry to hear that it isn't working @jaskinn. Please tell us more about your setup.

**Comment by jaskinn** [comment]
Running on Galaxy, 1.4.1.1.

Where should the 'onMigrate' overwrite occur?

**Comment by jaskinn** [comment]
So it rearranged how I was overwriting - I placed it in client/main.js inside a Meteor.startup() call. It does seem to have resolved the problem. Thank you @rjakobsson et al.

**Comment by abernix** [comment]
Just a shot in the dark, but is anyone who is still experiencing this problem using `meteorhacks:cluster`?

**Comment by TWKDreamState** [comment]
@abernix I can confirm that I am experiencing this issue, but I am not using meteorhacks:cluster.
@jaskinn @rjakobsson I did not have this in a Meteor.startup() call so I will try this tomorrow. Thank you for the info.

**Comment by laroygreen** [comment]
Hi all, 

I am also struggling with this issue. I tried adding `Meteor._reload.onMigrate(function () { return [false]; });` to my client/main.js file, but it did not work (and I know it is being called as I logged to console). As mentioned above, the browser infinitely refreshes and the page never loads.

I am using Meteor 1.4.1.1, deployed to a UNIX server (Ubuntu 16.04) running on NGINX 1.10.0. I have tested this on Chrome 53.0.2785.143 on the desktop and on Chrome Chrome 53.0.2785.124 on Android.

**Comment by PEM--** [comment]
I'm having the same issue when generating HTML files with an `Etag` and `Cache-Control` set to `public, no-cache`, and checking for `If-None-Match` for sending a 304.

For the moment, I circumvent it by using an analysis of the number of HTTP 304 that I send for a given IP and force an HTTP 200 in case this number is over 3 (I consider it a sign of a bad behavior of `Meteor.Reload`'s package).

Reproducing this isn't easy. But this pattern lead to more reproduction scenario of the `Meteor.Reload` bad behavior:

1. Create a `connect` middleware that:
    1. create an HTML payload with `dynamicHead` and `dynamicBody` depending on called routes,
    1. send `Etag` set to a hash of `dynamicHead` + `dynamicBody` and `Cache-Control` set to `public, no-cache`,
    1. use F5 on Chrome without its DevTools opened (or untick the option that clear the cache on each page reload),
    1. analyses `If-None-Match` and send 304 when `Etag` is matching.
1. Modify the server side part of your app without introducing new changes on the client side.

**Comment by sethjgore** [comment]
adding ```Meteor._reload.onMigrate(function () { return [false]; });```  in client/main.js worked for me.

**Comment by hwillson** [comment]
Reading over the issue thread here, and the long forum thread about this issue, there still isn't really anything concrete here that Meteor can help address. The solution mentioned [here](https://forums.meteor.com/t/app-constantly-refreshing-after-an-update/23586/124) seems to help, so that might be a first step for anyone new coming to this issue. 

Since there hasn't been any recent activity (with new findings) on this issue however (here or in the forums), and as part of the on-going old issue cleanup process, I'll close this issue for now. If anyone is still having this problem and can ideally help put together a reproduction that demonstrates this issue (which in this case I realize is quite difficult to do), that would really help with troubleshooting this further. If anyone thinks this issue should stay open, I'll consider re-opening it, but please bear in mind that keeping 1.5 year old issues open with no clear path forward doesn't really benefit anyone. If you want this issue to stay open, please provide additional findings that help define a way forward here. Thanks!

**Comment by dhrubins** [comment]
For us at Legal Robot, this happened constantly when we had an old version in our CDN cache. We wanted new versions of the app to push out with hot code push, so using the `_reload.onMigrate` solution didn't seem like a great option for us.

We use CloudFlare and a CI system to build and deploy new releases, so we resolved this by, at the beginning of the deployment process, just setting the cache into development mode to blow away the cache. We also lowered the time or set some of our cache headers to immediate expire.
